ENH: add NaiveRatioUniforms from UNU.RAN to scipy.stats

chrisb83 opened the pull request:
#### Reference issue
gh-14600

#### What does this implement/fix?
add NaiveRatioUniforms from UNU.RAN

this will replace `stats.rvs_ratio_uniforms` (a separate PR to deprecate that function will be submitted once this PR is merged)

#### Additional information

In line with `stats.rvs_ratio_uniforms`, the bounding rectangle has to be specified. One could make it optional and then UNU.RAN would try to find the boundaries numerically. I think it is better to ask users to compute the boundaries using `scipy.optimize` outside of this method to keep the interface simpler (for example, `mode` is not needed) and avoid additional test cases.

The tutorial still needs a bit more work: I will demonstrate how to use `r` and a mode shift (parameter `center`). The related test cases are already implemented, so I mainly need to explain the approach.

[ ] update tutorial
[ ] add benchmarks

Some general feedback on the approach that @tirthasheshpatel  implemented to wrap UNU.RAN: I found it very easy to use. Especially the functionality in `test_sampling.py` is very nice, e.g., to check invalid PDFs, domains etc

chrisb83 commented on benchmarks/benchmarks/stats_sampling.py:
```diff
@@ -181,6 +186,35 @@ def time_pinv_rvs(self, dist):
+            raise NotImplementedError(f"no rectangle specified for {dist}")
+
+    def time_nrou_setup(self, dist):
+        u_min, u_max, v_max = dist.nrou_bounds
```
this needs to be `dist.nrou_bounds()`, otherwise an error is raised when running the benchmarks

tirthasheshpatel replied:
also, if you decide to keep `u_min`, `u_max`, and `v_max` optional, we can also check the performance of setup when bounding rectangles are not available.

chrisb83 commented:
One small tweak of the benchmark code is needed. Lint failures in the files to generate the plots are easy to fix. I hope the rest is in good shape.

@tirthasheshpatel could you take a look at the PR when you have time?

tirthasheshpatel commented on scipy/stats/_unuran/unuran_wrapper.pyi:
```diff
@@ -92,6 +92,26 @@ class NumericalInversePolynomial(Method):
+                 *,
+                 domain: None | Tuple[float, float] = ...,
+                 r: float = ...,
+                 u_min: None | float = ...,
+                 u_max: None | float = ...,
+                 v_max: None | float = ...,
```
The constructor fails when any of these arguments is `None`:

```python
In [14]: rng = NaiveRatioUniforms(dist, center=0, r=1, u_min=None, u_max=None, v_max=None, random_state=urng)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-14-df7aad1c15e2> in <module>
----> 1 rng = NaiveRatioUniforms(dist, center=0, r=1, u_min=None, u_max=None, v_max=None, random_state=urng)

unuran_wrapper.pyx in scipy.stats._unuran.unuran_wrapper.NaiveRatioUniforms.__cinit__()

unuran_wrapper.pyx in scipy.stats._unuran.unuran_wrapper.NaiveRatioUniforms._validate_args()

TypeError: '<=' not supported between instances of 'NoneType' and 'int'
```

tirthasheshpatel commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1613,3 +1613,246 @@ cdef class DiscreteAliasUrn(Method):
+                  *,
+                  domain=None,
+                  r=1.0,
+                  u_min,
+                  u_max,
+                  v_max,
```
Maybe change this to:

```suggestion
                  u_min=None,
                  u_max=None,
                  v_max=None,
```

I checked locally and UNU.RAN automatically calculates these using numerical search if not given by the user. So, it would be nice to keep these parameters optional.

tirthasheshpatel commented on scipy/stats/tests/test_sampling.py:
```diff
@@ -835,3 +839,91 @@ def pdf(self, x):
+        return 0, np.inf
+
+
+class TestNaiveRatioUniforms:
```
I think the tests are already quite strong. If you want, you can also add the distributions from UNU.RAN's test file `t_nrou.c`. I looked at it and it tests with the `dist2` distribution present in `TransformedDensityRejection`'s test suite (on line 336 in `test_sampling.py`).

tirthasheshpatel commented on scipy/stats/tests/test_sampling.py:
```diff
@@ -835,3 +839,91 @@ def pdf(self, x):
+
+    def test_exceptions(self):
+        dist = StandardNormal()
+        # need vmin < vmax
```
```suggestion
        # need u_min < u_max
```

tirthasheshpatel commented on scipy/stats/tests/test_sampling.py:
```diff
@@ -835,3 +839,91 @@ def pdf(self, x):
+            NaiveRatioUniforms(dist, v_max=1, u_min=3, u_max=1)
+        with pytest.raises(ValueError, match=msg):
+            NaiveRatioUniforms(dist, v_max=1, u_min=1, u_max=1)
+        # need umax > 0
```
```suggestion
        # need v_max > 0
```

tirthasheshpatel commented on scipy/stats/tests/test_sampling.py:
```diff
@@ -835,3 +839,91 @@ def pdf(self, x):
+                                 random_state=7303)
+        check_cont_samples(rng, dist, (0, 1))
+
+    @pytest.mark.parametrize("p", [1.5, 2.83])
```
This test seems very comprehensive!

tirthasheshpatel reviewed (commented):
Sorry, it took me some time to get to this! I had an initial look at the code, docs, and tests. The tutorial is very nicely written! Tests also seem quite comprehensive. Some comments on the UI and tests. I will take a closer look at docs and tutorials later today.

tirthasheshpatel commented:
There is a significant difference between the performance of `NaiveRatioUniforms` and `rvs_ratio_uniforms`:

```py
In [1]: from scipy.stats import NaiveRatioUniforms, rvs_ratio_uniforms

In [2]: import numpy as np

In [3]: from math import exp

In [4]: class StandardNormal:
   ...:     def pdf(self, x):
   ...:         return exp(-0.5 * x*x)
   ...: 

In [5]: dist = StandardNormal()

In [6]: from math import sqrt

In [7]: u_bound = sqrt(dist.pdf(sqrt(2))) * sqrt(2)
   ...: v_max = sqrt(dist.pdf(0))

In [8]: urng = np.random.default_rng()

In [9]: rng = NaiveRatioUniforms(dist, center=0, r=1, u_min=-u_bound, u_max=u_bound, v_max=v_max, random_state=urng)

In [10]: %timeit rng.rvs(100_000)
121 ms ± 4.88 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [11]: class StandardNormal:
    ...:     def pdf(self, x):
    ...:         return np.exp(-0.5 * x*x)
    ...: 

In [12]: dist = StandardNormal()

In [13]: %timeit rvs_ratio_uniforms(dist.pdf, vmin=-u_bound, vmax=u_bound, umax=v_max, size=100_000, random_state=urng)
20 ms ± 1.2 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [14]:
```

UNU.RAN uses the same algorithm as `rvs_ratio_uniforms` to generate RVs. I think the slowdown is because `NaiveRatioUniforms` doesn't call the PDF in a vectorized manner like `rvs_ratio_uniforms` does. This makes me think if we should deprecate `rvs_ratio_uniforms`. Do you think this method is worth deprecating `rvs_ratio_uniforms`? Or should we keep both? If not, we could change `rvs_ratio_uniforms` into a class with a `rvs` method instead of using UNU.RAN. What do you think?

mdhaber commented on doc/source/tutorial/stats/sampling_nrou.rst:
```diff
@@ -0,0 +1,225 @@
+  * Set-up: fast (if the bounding rectangle is known), otherwise slow
+  * Sampling: moderate
+
+NROU is an implementation of the (generalized) ratio-of-uniforms method which
```
```suggestion
NROU is an special case of the (generalized) ratio-of-uniforms method that
```
Is that what is meant?

chrisb83 replied:
no, NROU is the usual ROU method if r==1, otherwise it is called generalized ROU

mdhaber commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1613,3 +1613,246 @@ cdef class DiscreteAliasUrn(Method):
+
+        * ``pdf``: PDF of the distribution. The signature of the PDF is
+          expected to be: ``def pdf(self, x: float) -> float``. i.e.
+          the PDF should accept a Python float and the parameters and
```
```suggestion
          the PDF should accept a Python float and
```

mdhaber commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1613,3 +1613,246 @@ cdef class DiscreteAliasUrn(Method):
+        Set parameter ``r > 0`` for generalized Ratio-Of-Uniforms method.
+        The default is 1.0.
+
+    u_min: float
```
```suggestion
    u_min: float, optional
```
(and below)? What does it do if left at the default, `None`?

mdhaber commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1613,3 +1613,246 @@ cdef class DiscreteAliasUrn(Method):
+        Sets the left boundary of the bounding rectangle, see details under
+        Notes. If necessary, consider computing the bound numerically using
+        methods in `scipy.optimize`. Important: The bound has to be
+        provided for ``dist.pdf(x - center)``.
```
Would it not be possible to allow the user to provide the bound for `x` rather than `x-center`? 
If so, I think we should discuss which would be the more natural interface.

chrisb83 replied:
`c` in `rvs_ratio_uniform` corresponds to `center`, the generalized version with r is not implemented in the old version

And yes, u and v have switched roles... UNU.RAN does not follow the notation introduced in the original article of Kindernan/Monahan. That is a bit confusing but the examples should help to understand the use of u and v

mdhaber commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1613,3 +1613,246 @@ cdef class DiscreteAliasUrn(Method):
+        then that instance is used.
+
+    Notes
+    -----
```
I won't comment specifically about uses of single/double backticks and how equations are represented, but please consider whether symbols are appearing as you intend.

mdhaber reviewed (commented):
Only looked at the interface, which we can't change. Two other things, which are things I didn't notice (or maybe just don't _remember_ discussing) in previous PRs:
* I'm not the big proponent of keyword-only args, but the idea is growing on me. I was asked to make everything after `pdf` keyword only in `NumericalInverseHermite`. Is `center` really so much more fundamental than `domain`? I suppose if we're going to have keyword-only args, I wonder why any of the optional parameters deserve to be not-keyword-only. If it's already been discussed, never mind.
* Does `set_random_state` work the same way as creating a new `NaiveRatioUniforms` with the same parameters and the new `random_state`, or is there a difference between that and only setting the `random_state` used by `rvs`?

mdhaber commented on scipy/stats/tests/test_sampling.py:
```diff
@@ -52,10 +54,12 @@ def support(self):
+    ("NumericalInversePolynomial", {"dist": StandardNormal()}),
+    ("NaiveRatioUniforms", {"dist": StandardNormal(),
+                            "u_min": -np.exp(-0.5) * np.sqrt(2),
+                            "u_max": np.exp(-0.5) * np.sqrt(2), "v_max": 1.0})
```
I don't know the status of UNU.RAN testing, but `NumericalInverseHermite` tests with all continuous distributions (minus a few exceptions). It looks like `NaiveRatioUniforms` will only test with `StandardNormal` here; is that right?

chrisb83 replied:
The reason is that you need to provide the bounds `u_min, ...`. if we make these parameters optional, one could run more test cases easily. But then it is mainly a test if the numerical optimization works :)

chrisb83 commented:
> I think the slowdown is because NaiveRatioUniforms doesn't call the PDF in a vectorized manner like rvs_ratio_uniforms does. 

Yes, I also noted the relatively slow performance in the CI benchmark. At this point, we cannot do much about it. Support of  `LowLevelCallable` should fix this. For now, it probably makes sense to mention the performance issue in the documentation and to keep `rvs_ratio_uniforms`.

chrisb83 commented:
> I suppose if we're going to have keyword-only args, I wonder why any of the optional parameters deserve to allow positional args. If it's already been discussed, never mind.

The idea was that `center` belongs to the distribution. The keyword only args belong to the generation method (in this sense, even domain belongs rather to the method since it can be changed to generated truncated variables of the same distribution). But maybe that is indeed not the right way. Tirth and I spoke about it yesterday and we thought it might be better to follow your suggestion.

tirthasheshpatel commented:
> Does `set_random_state` work the same way as creating a new `NaiveRatioUniforms` with the same parameters and the new `random_state`, or is there a difference between that and only setting the `random_state` used by `rvs`?

Yes, `set_random_state` works the same way as creating a new `NaiveRatioUniforms` with the same parameters and the new `random_state`. A `random_state` keyword in the `rvs` method has been introduced in gh-14703 for backward-compatibility reasons.  The current behavior is: If the user provides a non-none `random_state` argument in the `rvs` method, use that instead of the `random_state` passed in the constructor or the last call to `set_random_state`.

chrisb83 commented:
I hope I addressed all comments above. The main changes are:

- All parameters but `dist` are keyword only
- the parameters `u_min, u_max, v_max` are optional

tirthasheshpatel commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1613,3 +1614,276 @@ cdef class DiscreteAliasUrn(Method):
+        # the reason is that the numerical routine to find the bounds can only
+        # be used for both bounds at the same time
+        warn = False
+        if (u_min is None) and (u_max is not None):
+            u_max = None
+            warn = True
+        if (u_min is not None) and (u_max is None):
+            u_min = None
+            warn = True
+        if warn:
+            msg = ("Only of the values of u_min and u_max is None. "
+                   "The value that is not None is ignored and both values "
+                   "will be computed numerically.")
```
```suggestion
        if (u_min is None) ^ (u_max is None):
            u_max, u_min = None, None
            msg = ("Only of the values of u_min and u_max is None. "
                   "The value that is not None is ignored and both values "
                   "will be computed numerically.")
```

good place to use XOR :)

tirthasheshpatel reviewed (approved):
I looked at all the changes and they look good. The implementation seems correct. Tests are strong and comprehensive. The slowdown compared to `rvs_ratio_uniforms` is mentioned in the docs. Tutorial seems very comprehensive and nicely written. I don't think I have any other points except some small ones below.

One of the things to consider is the formatting of some parts of docs. For example, this looks very cluttered:

![image](https://user-images.githubusercontent.com/43181252/135727122-f7065799-7e7d-425d-991a-e7788835d38a.png)

We should probably use LaTeX instead.

Also, some parts are rendered as _italics_ while they should be rendered as ``inline code``. Like `area(pdf)` and `0.5 * area(pdf)` below:

![image](https://user-images.githubusercontent.com/43181252/135727164-3186e843-bb25-47a1-9be2-a110882eec30.png)

chrisb83 commented:
I was also thinking to use LaTeX. I ended up sticking to the inline code formatting: note that there are quite a few instances where we refer to parameters like pdf, center or expressions like area. Of course, one can write text in LaTeX, but I thought the benefits are not that strong and with the current version, one can recognize the parameters easily, e.g., in `v_max*(u_max - u_min)`. 

What I will definitely do is to get rid of the inconsistencies you pointed out and I think that A, R and (U,V) do not need to be in italics, just normal text should be fine. Would that be OK from your point of view?

tirthasheshpatel commented:
> What I will definitely do is to get rid of the inconsistencies you pointed out and I think that A, R and (U,V) do not need to be in italics, just normal text should be fine. Would that be OK from your point of view?

That would be very helpful.

chrisb83 commented:
> That would be very helpful.

Done!

tirthasheshpatel commented:
> > That would be very helpful.
> 
> Done!

Thanks, @chrisb83! I think @mdhaber's points have been addressed too. I will take a final look today and merge tomorrow if there are no new/blocking reviews.

chrisb83 commented:
> Commented on some typos. Everything else looks good!

Thanks for fixing the typos.
